src/main/java/io/cdap/plugin/salesforce/SObjectsDescribeResult.java

```diff
@@ -65,26 +65,60 @@ public SObjectsDescribeResult(PartnerConnection connection, Collection<String> s
   }
 
   /**
-   * Based in given list of parent SObject names, describes each SObject and stores describe results
-   * by full parent path name.
-   * <p/>
-   * For input [{Account=TOP_LEVEL}, {Account=TOP_LEVEL, Contacts=RELATIONSHIP}]
-   * describe result will contain description for: Account, Account.Contacts.
+   * Describes given top level SObject and its featured SObjects, i.e. SObjects
+   * which were present in the query due to parent-to-child or child-to-parent relationship usage.
    *
    * @param connection Salesforce partner connection
-   * @param parentSObjects unique combinations of parent SObject names and their types
+   * @param name top level SObject name
+   * @param featuredSObjects unique combination of featured SObjects
    */
-  public SObjectsDescribeResult(PartnerConnection connection,
-                                Set<Map<String, SObjectDescriptor.SObjectType>> parentSObjects) {
+  public SObjectsDescribeResult(PartnerConnection connection, String name,
+                                Collection<SObjectDescriptor.FeaturedSObjects> featuredSObjects) {
     Map<String, DescribeSObjectResult> cache = new HashMap<>();
-    parentSObjects
-      .forEach(names -> {
-        try {
-          addSObjectDescribe(connection, names, cache);
-        } catch (ConnectionException e) {
-          throw new RuntimeException(e);
+    try {
+      // describe top level SObject first and store it to be used during schema creation
+      DescribeSObjectResult topLevelDescribe = describe(connection, name, cache);
+      addSObjectDescribe(name, topLevelDescribe.getFields());
+
+      // describe each featured SObjects combination
+      for (SObjectDescriptor.FeaturedSObjects obj : featuredSObjects) {
+        // at the beginning last describe is top level SObject describe result
+        DescribeSObjectResult lastDescribe = topLevelDescribe;
+
+        /*
+          If featured SObjects have relationship, describe it first.
+          For the query `SELECT Name, (SELECT LastName, Owner.UserRole.Id FROM Contacts) FROM Account`
+          `Contacts` is relationship SObject.
+         */
+        if (obj.hasRelationship()) {
+          String relationshipName = getRelationshipName(obj.getRelationship(), lastDescribe);
+          if (relationshipName == null) {
+            throw new IllegalArgumentException("Unable to find relationship field name: " + name);
```

Comment:
I'm not sure if this error is descriptive enough. For the example above, 'name'=Account and obj.getRelationship()='Contacts' correct? Does this happen if 'Contacts' could not be found? If so a more detailed error message is needed.

Reply:
Updated error message.